lib/puppet/ssl/state_machine.rb

```diff
@@ -393,25 +393,43 @@ def ensure_client_certificate
   private
 
   def run_machine(state, stop)
-    with_lock do
-      loop do
-        state = run_step(state)
-
-        case state
-        when stop
-          break
-        when Error
-          if @onetime
-            Puppet.log_exception(state.error)
-            raise state.error
+    begin
+      with_lock do
+        loop do
+          state = run_step(state)
+
+          case state
+          when stop
+            break
+          when Error
+            if @onetime
+              Puppet.log_exception(state.error)
+              raise state.error
+            end
+          else
+            # fall through
           end
-        else
-          # fall through
         end
       end
-    end
 
-    state
+      state
+    rescue Puppet::LockError
+      wait_for_lock_deadline ||= Time.now.to_i + Puppet[:maxwaitforlock]
+
+      if Puppet[:maxwaitforlock] < 1
+        Puppet.notice _("Run of %{client_class} already in progress; skipping  (%{lockfile_path} exists, and waitforlock setting is set to 0)") % { client_class: client_class, lockfile_path: lockfile_path }
+        exit(1)
+      elsif Time.now.to_i > wait_for_lock_deadline
+        Puppet.notice _("Exiting now because the maxwaitforlock timeout has been exceeded.")
+        exit(1)
```

Comment:
it's surprising to me to exit from within the state machine. I'd expect this to raise an exception so that the application (`puppet agent` or `puppet ssl`) can cleanup, log the error, etc.